### DISCUSSION TITLE
Techniques to smooth 3D mesh from labeled image

### DISCUSSION THREAD

**Discussion by oarcelus** [CGAL/cgal#8285] [General]
When trying to build a 3D tetrahedral mesh from labeled image I use the examples shown in the webpage (https://doc.cgal.org/latest/Mesh_3/index.html)

The meshes I am dealing with are quite complex with pretty amorphous shapes. The issue I am having is that the domain boundaries sometimes come up pretty 'spiky'. Some tetrahedra go into other domains instead of being bound to their surface.
![MicrosoftTeams-image](https://github.com/CGAL/cgal/assets/24278825/b1e61afb-a0cc-48a9-bf62-5ffe330b8288)

I guess that one way of making these shapes smoother is to increase the resolution of the underlying 3D image (make smaller voxels). But I wonder if there is some better way to inprove this withing CGAL.

Best regards,
Oier.

**Comment by janetournois** [comment]
Hello,

you should try meshing from weighted images, as done in [this example](https://doc.cgal.org/6.0-beta1/Mesh_3/Mesh_3_2mesh_3D_weighted_image_8cpp-example.html)

Then, reducing your `facet_distance` criterion should help.

Best regards,
Jane

**Reply by oarcelus** [reply]
Thanks! This improves a little bit. I just discovered that the reason these spikes appear is because I am also meshing the surrounding 'air' volume for my simulations. If I exclude the surrounding volume from the meshing, my particles are much smoother.

So, I wonder if there is a way of first meshing the particles alone (even individually since I have labels for each of them), and then keep these surface polyhedral features to mesh the whole structure.

I have tried to include feature detection as a partial solution, to at least improve the meshin in the contact between 3 subdomains. But I find some errors when compiling.

    CGAL::Image_3 image;
      const bool success = image.read(inr_filename.c_str());
      if (!success) {
      throw "Could not read image file";
      }
    
      // Create weights with gaussian smoothing with sigma = sig * max_voxel_size
      const float sigma = sig * (std::max)(image.vx(), (std::max)(image.vy(), image.vz()));
      CGAL::Image_3 img_weights = CGAL::Mesh_3::generate_label_weights(image, sigma);
    
      Mesh_domain cgal_domain = Mesh_domain::create_labeled_image_mesh_domain(image,
                                                                              CGAL::parameters::weights(img_weights).relative_error_bound(relative_error_bound_weights),
                                                                              CGAL::parameters::features_detector = CGAL::Mesh_3::Detect_features_in_image());

I see an error of the type:
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Labeled_mesh_domain_3.h(848,66): error C2672: 'CGAL::internal_np::combine_named_parameters': no matching overloaded funct
ion found [C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\mesh_from_labeled_3d_array_with_weights_detect.cpp(76): message : see reference to function template instantiation 'auto
CGAL::Labeled_mesh_domain_3<K,int,std::pair<int,int>>::create_labeled_image_mesh_domain<double,CGAL::internal_np::error_bound_t,CGAL::Named_function_parameters<std::refe
rence_wrapper<CGAL::Image_3>,Tag,CGAL::internal_np::No_property>,CGAL::Mesh_3::Detect_features_in_image,CGAL::internal_np::features_detector_param_t,CGAL::internal_np::N
o_property,>(const CGAL::Image_3 &,const CGAL::Named_function_parameters<double,CGAL::internal_np::error_bound_t,CGAL::Named_function_parameters<std::reference_wrapper<C
GAL::Image_3>,Tag,CGAL::internal_np::No_property>> &,const CGAL::Named_function_parameters<CGAL::Mesh_3::Detect_features_in_image,CGAL::internal_np::features_detector_pa
ram_t,CGAL::internal_np::No_property> &)' being compiled [C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
          with
          [
              Tag=CGAL::internal_np::weights_param_t
          ]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Labeled_mesh_domain_3.h(847,1): error C2784: 'auto CGAL::internal_np::combine_named_parameters(const CGAL::Named_function
_parameters<K,Tag,CGAL::internal_np::No_property> &,const NPS &...)': could not deduce template argument for 'const CGAL::Named_function_parameters<K,Tag,CGAL::internal_
np::No_property> &' from 'const CGAL::Named_function_parameters<double,CGAL::internal_np::error_bound_t,CGAL::Named_function_parameters<std::reference_wrapper<CGAL::Imag
e_3>,Tag,CGAL::internal_np::No_property>>' [C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
          with
          [
              Tag=CGAL::internal_np::weights_param_t
          ]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Named_function_parameters.h(329): message : see declaration of 'CGAL::internal_np::combine_named_parameters' [C:\Users\oa
rcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Labeled_mesh_domain_3.h(847,1): error C2782: 'auto CGAL::internal_np::combine_named_parameters(const CGAL::Named_function
_parameters<K,Tag,CGAL::internal_np::No_property> &,const NPS &...)': template parameter 'K' is ambiguous [C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_fro
m_labeled_3d_array_with_weights_detect.vcxproj]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Named_function_parameters.h(329): message : see declaration of 'CGAL::internal_np::combine_named_parameters' [C:\Users\oa
rcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Labeled_mesh_domain_3.h(847,1): message : could be 'std::reference_wrapper<CGAL::Image_3>' [C:\Users\oarcelus\ElectrodeBu
ilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]
C:\dev\vcpkg\installed\x64-windows\include\CGAL/Labeled_mesh_domain_3.h(847,1): message : or       'double' [C:\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_f
rom_labeled_3d_array_with_weights_detect.vcxproj]

**Comment by janetournois** [comment]
You probably can't mix the named parameters `,` and `.` separators together.
Try
```
esh_domain cgal_domain = Mesh_domain::create_labeled_image_mesh_domain(image,
                                                                          CGAL::parameters::weights(img_weights).relative_error_bound(relative_error_bound_weights)
                                                                          .features_detector(CGAL::Mesh_3::Detect_features_in_image()));
 ```

**Reply by oarcelus** [reply]
Hi, I now get a linking error.

mesh_from_labeled_3d_array_with_weights_detect.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) v
oid __cdecl itk::OutputWindowDisplayWarningText(char const *)" (__imp_?OutputWindowDisplayWarningText@itk@@YAXPEBD@Z) r
eferenced in function "protected: virtual class std::vector<double,class std::allocator<double> > __cdecl itk::Gaussian
Operator<unsigned char,3,class itk::NeighborhoodAllocator<unsigned char> >::GenerateCoefficients(void)" (?GenerateCoeff
icients@?$GaussianOperator@E$02V?$NeighborhoodAllocator@E@itk@@@itk@@MEAA?AV?$vector@NV?$allocator@N@std@@@std@@XZ) [C:
\Users\oarcelus\ElectrodeBuilder\meshing_cgal\build\mesh_from_labeled_3d_array_with_weights_detect.vcxproj]

It is strange because when I include feature detection only or weights only it works. But combining both not. Any idea how could I keep advancing here?

Best regardsm
Oier.

**Reply by janetournois** [reply]
Hello,

You can have a look at the example [mesh_3D_weighted_image_with_detection_of_features.cpp](https://github.com/CGAL/cgal/blob/master/Mesh_3/examples/Mesh_3/mesh_3D_weighted_image_with_detection_of_features.cpp) which combines weights and features detection as you are trying to do.

The [corresponding CMakeLists](https://github.com/CGAL/cgal/blob/master/Mesh_3/examples/Mesh_3/CMakeLists.txt) prepares the proper setup for all examples of Mesh_3, including this one.

**Reply by oarcelus** [reply]
Oh, there was an example all along. I did not see that! Thanks!

**Reply by janetournois** [reply]
The example was there but not listed in the documentation.
The PR ##8373 adds it to the doc
